File: golang/leetcode/array-string/find-the-index-of-the-first-occurrence-in-a-string.go
package main

import (
	"fmt"
)

func strStr(haystack string, needle string) int {
	n, m := len(haystack), len(needle)

	if haystack == needle {
		return 0
	}

	for i := 0; i <= n-m; i++ {
		if haystack[i:i+m] == needle {
			return i
		}
	}

	return -1
}

func main() {
	haystack1 := "sadbutsad"
	needle1 := "sad"
	fmt.Println(strStr(haystack1, needle1)) // Output: 0

	haystack2 := "leetcode"
	needle2 := "leeto"
	fmt.Println(strStr(haystack2, needle2)) // Output: -1
}

File: python/leetcode/array-string/find-the-index-of-the-first-occurrence-in-a-string.py
# 28. Find the Index of the First Occurrence in a String
def strStr(haystack, needle):
    n, m = len(haystack), len(needle)

    if m == 0:
        return 0

    for i in range(n - m + 1):
        if haystack[i:i + m] == needle:
            return i

    return -1

def strStr_v2(haystack, needle):
    n, m = len(haystack), len(needle)

    if m == 0:
        return 0

    i = 0
    while i <= n - m:
        j = 0
        while j < m and haystack[i + j] == needle[j]:
            j += 1
        if j == m:
            return i
        i += 1

    return -1


if __name__ == '__main__':
    haystack = "sadbutsad"
    needle = "sad"

    print(strStr_v2(haystack, needle))  # Output: 0

    haystack = "leetcode"
    needle = "leeto"

    print(strStr_v2(haystack, needle))  # Output: -1
